### PR TITLE
Implementation ANS102: bundle data 

### PR DESCRIPTION
**Pull request by zyjblockchain**

### CODE REVIEW

**Comment by samsends** [comment]
hey @outprog how's this looking? We'd like to start using bundles

**Comment by zyjblockchain** [comment]
> hey @outprog how's this looking? We'd like to start using bundles

We need to test it and merge.
It will probably be online tomorrow.

File: bundle.go [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+	bdBinary = append(bdBinary, utils.LongTo32ByteArray(len(dataItems))...)
+	bdBinary = append(bdBinary, headers...)
+	bdBinary = append(bdBinary, binaries...)
+	return &BundleData{
+		Items:        dataItems,
+		bundleBinary: bdBinary,
+	}, nil
+}
+
+func RecoverBundleData(bundleBinary []byte) (*BundleData, error) {
```
Comment (outprog) [inline bundle.go:46]: DecodeBundleDate or ParseBundleDate?

File: wallet.go [hunk 2]
```diff
@@ -147,3 +148,91 @@ func (w *Wallet) SendTransaction(tx *types.Transaction) (id string, err error) {
 	}
 	return
 }
+
+// about bundle tx
+
+func (w *Wallet) CreateBundleDataItem(data []byte, signatureType int, target string, anchor string, tags []types.Tag) (di DataItem, err error) {
```
Comment (outprog) [inline wallet.go:154]: Should be split
bundles := CreateBundleDateItem()
e := w.SignBundle(bundles)

Reference: utils.SignTransaction

File: types/bundle.go [hunk 3]
```diff
@@ -0,0 +1,19 @@
+package types
+
+type BundleData struct {
+	Items        []DataItem `json:"items"`
+	BundleBinary []byte
+}
+
+type DataItem struct {
```
Comment (outprog) [inline types/bundle.go:8]: rename to BundleItem ？

File: types/const.go [hunk 4]
```diff
@@ -42,3 +42,9 @@ var FATAL_CHUNK_UPLOAD_ERRORS = map[string]struct{}{
 	"chunk_proof_ratio_not_attractive": struct{}{},
 	"invalid_proof":                    struct{}{},
 }
+
+// about bundle
+const (
+	BUNDLER         = "http://bundler.arweave.net:10000"
```
Comment (outprog) [inline types/const.go:48]: rename to BUNDLER_URL ?

File: types/const.go [hunk 5]
```diff
@@ -42,3 +42,9 @@ var FATAL_CHUNK_UPLOAD_ERRORS = map[string]struct{}{
...
+	MIN_BINARY_SIZE = 1044
```
Comment (outprog) [inline types/const.go:49]: MIN_BUNDLE_SIZE?

File: utils/bundle.go [hunk 6]
```diff
@@ -0,0 +1,325 @@
...
+		if dataItem.Id != id {
+			return nil, fmt.Errorf("dataItem.Id != id, dataItem.Id: %s, id: %s", dataItem.Id, id)
+		}
+		bd.Items = append(bd.Items, *dataItem)
+		dataItemStart += itemBinaryLength
+	}
+	return bd, nil
+}
+
+func DecodeDataItem(itemBinary []byte) (*types.DataItem, error) {
```
Comment (outprog) [inline utils/bundle.go:81]: DecodeBundleItem

File: utils/bundle.go [hunk 7]
```diff
@@ -0,0 +1,325 @@
...
+		Target:        target,
+		Anchor:        anchor,
+		Tags:          tags,
+		Data:          Base64Encode(data),
+		Id:            id,
+		ItemBinary:    itemBinary,
+	}, nil
+}
+
+func NewDataItem(owner, signatureType, target, anchor string, data []byte, tags []types.Tag) *types.DataItem {
```
Comment (outprog) [inline utils/bundle.go:135]: BundleItem

**Comment by outprog** [comment]
> hey @outprog how's this looking? We'd like to start using bundles

Hey @samsends bundle is done. v1.3.0 https://github.com/everFinance/goar#createbundle

**Pull request merged**